File: src/main/java/au/id/micolous/metrodroid/transit/china/NewShenzhenTransitData.java
/*
 * NewShenzhenTransitData.java
 *
 * Copyright 2018 Google
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package au.id.micolous.metrodroid.transit.china;

import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import au.id.micolous.metrodroid.multi.Localizer;
import au.id.micolous.metrodroid.util.NumberUtils;
import org.jetbrains.annotations.NonNls;

import java.util.Collections;
import java.util.List;

import au.id.micolous.farebot.R;
import au.id.micolous.metrodroid.card.CardType;
import au.id.micolous.metrodroid.card.china.ChinaCardTransitFactory;
import au.id.micolous.metrodroid.card.iso7816.ISO7816File;
import au.id.micolous.metrodroid.card.china.ChinaCard;
import au.id.micolous.metrodroid.card.iso7816.ISO7816TLV;
import au.id.micolous.metrodroid.transit.CardInfo;
import au.id.micolous.metrodroid.transit.TransitData;
import au.id.micolous.metrodroid.transit.TransitIdentity;
import au.id.micolous.metrodroid.util.Utils;
import au.id.micolous.metrodroid.xml.ImmutableByteArray;

// Reference: https://github.com/sinpolib/nfcard/blob/master/src/com/sinpo/xnfc/nfc/reader/pboc/ShenzhenTong.java
public class NewShenzhenTransitData extends ChinaTransitData {
    private final int mSerial;

    public static final CardInfo CARD_INFO = new CardInfo.Builder()
            .setImageId(R.drawable.szt_card)
            .setName(Localizer.INSTANCE.localizeString(R.string.card_name_szt))
            .setLocation(R.string.location_shenzhen)
            .setCardType(CardType.FeliCa)
            .setPreview()
            .build();

    public static final Parcelable.Creator<NewShenzhenTransitData> CREATOR = new Parcelable.Creator<NewShenzhenTransitData>() {
        public NewShenzhenTransitData createFromParcel(Parcel parcel) {
            return new NewShenzhenTransitData(parcel);
        }

        public NewShenzhenTransitData[] newArray(int size) {
            return new NewShenzhenTransitData[size];
        }
    };

    private NewShenzhenTransitData(ChinaCard card) {
        super(card);
        mSerial = parseSerial(card);
        ImmutableByteArray szttag = getTagInfo(card);

        if (szttag != null) {
            mValidityStart = szttag.byteArrayToInt(20, 4);
            mValidityEnd = szttag.byteArrayToInt(24, 4);
        }
    }

    @Override
    protected ChinaTrip parseTrip(ImmutableByteArray data) {
        return new NewShenzhenTrip(data);
    }

    @Override
    public String getSerialNumber() {
        return formatSerial(mSerial);
    }

    @Override
    public String getCardName() {
        return Localizer.INSTANCE.localizeString(R.string.card_name_szt);
    }

    @Override
    public void writeToParcel(Parcel parcel, int flags) {
        super.writeToParcel(parcel, flags);
        parcel.writeInt(mSerial);
    }

    private NewShenzhenTransitData(Parcel parcel) {
        super(parcel);
        mSerial = parcel.readInt();
    }

    public final static ChinaCardTransitFactory FACTORY = new ChinaCardTransitFactory() {
        @Override
        public List<ImmutableByteArray> getAppNames() {
            return Collections.singletonList(ImmutableByteArray.Companion.fromASCII("PAY.SZT"));
        }

        @Override
        public TransitIdentity parseTransitIdentity(@NonNull ChinaCard card) {
            return new TransitIdentity(Localizer.INSTANCE.localizeString(R.string.card_name_szt), formatSerial(parseSerial(card)));
        }

        @Override
        public TransitData parseTransitData(@NonNull ChinaCard chinaCard) {
            return new NewShenzhenTransitData(chinaCard);
        }

        @NonNull
        @Override
        public List<CardInfo> getAllCards() {
            return Collections.singletonList(CARD_INFO);
        }
    };

    @NonNls
    private static String formatSerial(int sn) {
        int digsum = NumberUtils.INSTANCE.getDigitSum(sn);
        // Sum of digits must be divisible by 10
        int lastDigit = (10 - (digsum % 10)) % 10;
        return Integer.toString(sn) + "(" + Integer.toString(lastDigit) + ")";
    }

    @Nullable
    private static ImmutableByteArray getTagInfo(ChinaCard card) {
        ISO7816File file15 = getFile(card, 0x15);
        if (file15 != null)
            return file15.getBinaryData();
        ImmutableByteArray szttag = ISO7816TLV.INSTANCE.findBERTLV(card.getAppData(), "a5", true);
        if (szttag == null)
            return null;
        return ISO7816TLV.INSTANCE.findBERTLV(szttag, "8c", false);
    }

    private static int parseSerial(ChinaCard card) {
        ImmutableByteArray ti = getTagInfo(card);
        if (ti == null)
            return 0;
        return ti.byteArrayToIntReversed(16,4);
    }
}

File: src/commonMain/kotlin/au/id/micolous/metrodroid/transit/china/NewShenzhenTransitData.kt
/*
 * NewShenzhenTransitData.kt
 *
 * Copyright 2018 Google
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package au.id.micolous.metrodroid.transit.china

import au.id.micolous.metrodroid.card.CardType
import au.id.micolous.metrodroid.card.china.ChinaCardTransitFactory
import au.id.micolous.metrodroid.card.china.ChinaCard
import au.id.micolous.metrodroid.card.iso7816.ISO7816TLV
import au.id.micolous.metrodroid.multi.Localizer
import au.id.micolous.metrodroid.multi.Parcelize
import au.id.micolous.metrodroid.multi.R
import au.id.micolous.metrodroid.transit.*
import au.id.micolous.metrodroid.util.NumberUtils
import au.id.micolous.metrodroid.util.ImmutableByteArray

// Reference: https://github.com/sinpolib/nfcard/blob/master/src/com/sinpo/xnfc/nfc/reader/pboc/ShenzhenTong.java
@Parcelize
class NewShenzhenTransitData (val validityStart: Int?,
                              val validityEnd: Int?,
                              private val mSerial: Int,
                              override val trips: List<NewShenzhenTrip>?,
                              val mBalance: Int?): TransitData() {

    override val serialNumber: String?
        get() = formatSerial(mSerial)

    override val cardName: String
        get() = Localizer.localizeString(R.string.card_name_szt)

    public override val balance: TransitBalance?
        get() = if (mBalance != null)
            TransitBalanceStored(TransitCurrency.CNY(mBalance),
                    null,
                    ChinaTransitData.parseHexDate(validityStart),
                    ChinaTransitData.parseHexDate(validityEnd))
        else
            null

    companion object {
        private fun parse(card: ChinaCard) : NewShenzhenTransitData {
            val szttag = getTagInfo(card)

            return NewShenzhenTransitData(
                    validityStart = szttag?.byteArrayToInt(20, 4),
                    validityEnd = szttag?.byteArrayToInt(24, 4),
                    trips = ChinaTransitData.parseTrips(card)  { data -> NewShenzhenTrip(data) },
                    mSerial = parseSerial(card),
                    mBalance = ChinaTransitData.parseBalance(card)
            )
        }

        val CARD_INFO = CardInfo(
                imageId = R.drawable.szt_card,
                name = R.string.card_name_szt,
                locationId = R.string.location_shenzhen,
                cardType = CardType.FeliCa,
                region = TransitRegion.CHINA,
                preview = true)

        val FACTORY: ChinaCardTransitFactory = object : ChinaCardTransitFactory {
            override val appNames: List<ImmutableByteArray>
                get() = listOf(ImmutableByteArray.fromASCII("PAY.SZT"))

            override val allCards: List<CardInfo>
                get() = listOf(CARD_INFO)

            override fun parseTransitIdentity(card: ChinaCard) =
                    TransitIdentity(Localizer.localizeString(R.string.card_name_szt), formatSerial(parseSerial(card)))

            override fun parseTransitData(card: ChinaCard) = parse(card)
        }

        private fun formatSerial(sn: Int): String {
            val digsum = NumberUtils.getDigitSum(sn.toLong())
            // Sum of digits must be divisible by 10
            val lastDigit = (10 - digsum % 10) % 10
            return "$sn($lastDigit)"
        }

        private fun getTagInfo(card: ChinaCard): ImmutableByteArray? {
            val file15 = ChinaTransitData.getFile(card, 0x15)
            if (file15 != null)
                return file15.binaryData
            val szttag = card.appProprietaryBerTlv ?: return null
            return ISO7816TLV.findBERTLV(szttag, "8c", false)
        }

        private fun parseSerial(card: ChinaCard) = getTagInfo(card)?.byteArrayToIntReversed(16, 4) ?: 0
    }
}
